### PR TITLE
TIP-694: simple and stupid implementation of the completeness generator

### PR DESCRIPTION
**Pull request by jjanvier**
## What's here?

This PR extracts `schedule*` methods of the `CompletenessGeneratorInterface` into a new `CompletenessRemoverInterface`. The idea is to separate the concerns between the generation of the completenesses and the their removal. 

The PR also implements a very stupid and simple object version of the completeness manager. I chose to use exactly the same signature than what we have before to be sure the application works in a similar way than what we had. Contrary to the former implementation, this one is done in memory.

Those changes allowed to clean the changes we have made in the `ProductSaver` on this branch. Now it's really simple, it doesn't have any completeness calculation stuff anymore and doesn't have any SQL queries anymore.

To be able to know which completenesses should be recalculated, I added an `IS_EMPTY` filter on the completeness.

The completeness is emptied when:
* the family requirements change
* a locale is deleted or added to a channel
* and a lot of others use cases ^^

When the completenesses are emptied, products need to be reindexed.

## Not perfect, but ISO with what we have

Honestly, this behavior is really not perfect. A lot of functional use cases are not handled, just like now. See this Confluence page https://akeneo.atlassian.net/wiki/display/AKN/Completeness+calculation.

With this PR, completenesses removal is slower than before. We don't really have the choice right now as we need to reindex products.

The real solution would to recalculate the completeness via a batch job everytime needed. This would have the following benefits:
- no more heavy synchronous processes (yes, we have some...)
- users would be notified when things are done
- a single entry point for all the different use cases
- not that important if it's slower than before

But this is not the scope of this PR :)

## todo
- [x] IS_EMPTY operator on completeness
- [x] doc about completeness filter
- [x] revert Damien's commit about completeness manager
- [x] simple completeness generator

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | Y

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Resources/config/completeness.yml [hunk 1]
```diff
@@ -9,3 +10,12 @@ services:
...
             - '@pim_catalog.repository.cached_locale'
             - '@pim_catalog.completeness.checker'
+
+    pim_catalog.completeness.generator:
+        class: '%pim_catalog.completeness.generator.class%'
+        arguments:
+            - '@pim_catalog.query.product_query_builder_factory'
+            - '@pim_catalog.completeness.calculator'
+            - '@pim_catalog.saver.product'
+            - '@pim_catalog.saver.product'
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Resources/config/completeness.yml]: Well there is a problem here

**Review by SamirBoulil — DISMISSED** [review]

File: src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        $this->productBulkSaver = $productBulkSaver;
+        $this->completenessRemover = $completenessRemover;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateMissingForProduct(ProductInterface $product)
+    {
+        $this->completenessCalculator->calculate($product);
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php]: Useless :)

File: src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        $this->productSaver->save($product);
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php]: Shouldn't you instead move the logic from the ProductSaver to here, so the ProductSaver use just the completeness manager like before?

File: src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        $products = $this->createProductQueryBuilderForMissings()->execute();
+        $this->calculateAndBulkSave($products);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function schedule(ProductInterface $product)
+    {
+        $product->getCompletenesses()->clear();
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php]: Same comment, the saver already do a clear, but it should not.

File: src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    public function generateMissing()
+    {
+        $products = $this->createProductQueryBuilderForMissings()->execute();
+        $this->calculateAndBulkSave($products);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function schedule(ProductInterface $product)
```
Comment (damien-carcel) [inline src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php]: I know it is to keep the old name, but this isn't a schedule anymore. `schedule` was only to clear, not to calculate.

Comment (jjanvier) [reply src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php]: here is does not calculate at all
I didn't change the way things work
I'm ISO with what we had

Comment (damien-carcel) [reply src/Pim/Component/Catalog/Completeness/CompletenessGenerator.php]: Indeed, sorry, I misread.

**Review by damien-carcel — DISMISSED** [review]

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+use Pim\Component\Catalog\Model\FamilyInterface;
+use Pim\Component\Catalog\Model\LocaleInterface;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderFactoryInterface;
+use Pim\Component\Catalog\Query\ProductQueryBuilderInterface;
+
+/**
+ * Simple ORM version of the completeness remover.
+ * Please not that completenesses are also removed from the index.
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php]: note

**Review by damien-carcel — APPROVED** [review]

**Review by SamirBoulil — APPROVED** [review]

File: CHANGELOG-1.8.md [hunk 7]
```diff
@@ -12,6 +12,7 @@
 
 ### Constructors
 
+- Extract methods `schedule*` of `Pim\Component\Catalog\Completeness\CompletenessGeneratorInterface` into a `Pim\Component\Catalog\Completeness\CompletenessRemoverInterface`. Methods `schedule`, `scheduleForFamily` and `scheduleForChannelAndLocale` have been renamed respectively `removeForProduct`, `removeForFamily` and `removeForChannelAndLocale`.
```
Comment (BitOne) [inline CHANGELOG-1.8.md:15]: I'm not sure about the remove. I think **erase** or **delete** is somewhat more clear. Remove may imply that is has been removed from the product but can still exist outside of it. **drop** would work too.

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+use Pim\Component\Catalog\Model\ProductInterface;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderFactoryInterface;
+use Pim\Component\Catalog\Query\ProductQueryBuilderInterface;
+
+/**
+ * Simple ORM version of the completeness remover.
+ * Please note that completenesses are also removed from the index.
+ *
+ * @author    Julien Janvier (j.janvier@akeneo.com)
```
Comment (BitOne) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php]: Not sure this email exists ;) Maybe your IDE is not configured properly ? :p

File: CHANGELOG-1.8.md (hunk 7 above)
Comment (jjanvier) [reply CHANGELOG-1.8.md:15]: seen IRL, we'll keep as is, as we already have some "Remover" classes

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php (hunk 8 above)
Comment (jjanvier) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/CompletenessRemover.php]: fixed

**Comment by jjanvier** [comment]
we introduce another type of normalizer here
and we'll have to clean our normalizer stack as it's becoming quite huge

we discussed it with Marie and agreed that:
- we won't do it in this PR
- we'll do it for the 1.8 (dunno when exactly)

**Pull request merged**